main/CommandHandler.cpp

```diff
@@ -989,6 +989,200 @@ int setAnalogWrite(const uint8_t command[], uint8_t response[])
+
+  int ret = -1;
+  struct stat st;
+  if (stat(filename, &st) == 0) {
+    // Delete it if it exists
+    ret = unlink(filename);
+  }
+  return 0;
+}
+
+#include <driver/uart.h>
+
+int applyOTA(const uint8_t command[], uint8_t response[]) {
+#ifdef UNO_WIFI_REV2
+
+  const char* filename = "/fs/UPDATE.BIN";
+  FILE* updateFile = fopen(filename, "rb");
+
+  // init uart and write update to 4809
+  uart_config_t uart_config;
+
+  uart_config.baud_rate = 115200;
+  uart_config.data_bits = UART_DATA_8_BITS;
+  uart_config.parity = UART_PARITY_DISABLE;
+  uart_config.stop_bits = UART_STOP_BITS_1;
+  uart_config.flow_ctrl = UART_HW_FLOWCTRL_DISABLE;
+  uart_config.rx_flow_ctrl_thresh = 122;
+  uart_config.use_ref_tick = true;
+
+  uart_param_config(UART_NUM_1, &uart_config);
+
+  uart_set_pin(UART_NUM_1,
+                 1, // tx
+                 3, // rx
+                 UART_PIN_NO_CHANGE, // rts
+                 UART_PIN_NO_CHANGE); //cts
+
+  uart_driver_install(UART_NUM_1, 1024, 0, 20, NULL, 0);
+
+  struct stat st;
+  stat(filename, &st);
+
+  int retries = 0;
+
+  size_t remaining =  st.st_size % 1024;
+  for (int i=0; i<st.st_size; i++) {
+    uint8_t c;
+    uint8_t d;
+
+    fread(&c, 1, 1, updateFile);
+    retries = 0;
+    while (retries == 0 || (c != d && retries < 100)) {
+      uart_write_bytes(UART_NUM_1, (const char*)&c, 1);
+      uart_read_bytes(UART_NUM_1, &d, 1, 10);
+      retries++;
+    }
+    if (retries >= 100) {
+      goto exit;
+    }
+  }
+  // send remaining bytes (to reach page size) as 0xFF
+  for (int i=0; i<remaining + 10; i++) {
+    uint8_t c = 0xFF;
+    uint8_t d;
+    retries = 0;
+    while (retries == 0 || (c != d && retries < 100)) {
+      uart_write_bytes(UART_NUM_1, (const char*)&c, 1);
+      uart_read_bytes(UART_NUM_1, &d, 1, 10);
+      retries++;
+    }
+  }
+
+  // delay a bit before restarting, in case the flashing isn't yet over
+  delay(200);
+
+  pinMode(19, OUTPUT);
+  digitalWrite(19, HIGH);
+  delay(200);
+  digitalWrite(19, LOW);
+  pinMode(19, INPUT);
+
+exit:
+  fclose(updateFile);
+  unlink(filename);
+
+  return 0;
+#else
+  return 0;
+#endif
+}
+
+int existsFile(const uint8_t command[], uint8_t response[]) {
+  char filename[32 + 1];
+  size_t len;
+  size_t offset;
+
+  memcpy(&offset, &command[4], command[3]);
+  memcpy(&len, &command[5 + command[3]], command[4 + command[3]]);
+
+  memset(filename, 0x00, sizeof(filename));
+  memcpy(filename, &command[6 + command[3] + command[4 + command[3]]], command[5 + command[3] + command[4 + command[3]]]);
+
+  int ret = -1;
+
+  struct stat st;
+  ret = stat(filename, &st);
+  if (ret != 0) {
+    st.st_size = -1;
+  }
+  memcpy(&response[4], &(st.st_size), sizeof(st.st_size));
+
+  response[2] = 1; // number of parameters
+  response[3] = sizeof(st.st_size); // parameter 1 length
+
+  return 10;
+}
+
+int downloadFile(const uint8_t command[], uint8_t response[]) {
```

Comment:
I might be missing a piece in `arduinoOTA` but is this needed? From what I remember the arduinoOTA binary POSTs to a web server ...

Reply:
I added that method to allow a board to perform OTA even if it's behind a NAT (mostly for Create :slightly_smiling_face: ). I used it also for local deployment 
to reduce the amount of code that must be changed to adapt WiFi101OTA for UNO WiFi r2